src/core/BigDecimal.ts

```diff
@@ -0,0 +1,326 @@
+        if (val < BigInt(0)) {
+            return val.valueOf() * BigInt(-1);
+        } else {
+            return val;
+        }
+    }
+
+    /** @internal */
+    private signum(): number {
+        return this.unscaledValue > BigInt(0) ? 1 : (this.unscaledValue < BigInt(0) ? -1 : 0);
+    }
+
+    /**
+     * Returns a string representation of this `BigDecimal`
+     * without an exponent field.  For values with a positive scale,
+     * the number of digits to the right of the decimal point is used
+     * to indicate scale. For values with a zero or negative scale,
+     * the resulting string is generated as if the value were
+     * converted to a numerically equal value with zero scale and as
+     * if all the trailing zeros of the zero scale value were present
+     * in the result.
+     *
+     * The entire string is prefixed by a minus sign character '-'
+     * (<code>'&#92;u002D'</code>) if the unscaled value is less than
+     * zero. No sign character is prefixed if the unscaled value is
+     * zero or positive.
+     *
+     * @return a string representation of this `BigDecimal`
+     * without an exponent field.
+     */
+    toString(): string {
+        if (this.scale === 0) {
+            return this.unscaledValue.toString();
+        }
+        if (this.scale < 0) {
+            if (this.signum() === 0) {
+                return '0';
+            }
+            const trailingZeros = -this.scale;
+            let buf = '';
+            buf += this.unscaledValue.toString();
+            for (let i = 0; i < trailingZeros; i++) {
+                buf += '0';
+            }
+            return buf;
+        }
+        const str = BigDecimal.bigIntAbs(this.unscaledValue).toString();
+        return BigDecimal.getValueString(this.signum(), str, this.scale);
+    }
+
+    /** @internal */
+    private static getValueString(signum: number, intString: string, scale: number): string {
+        /* Insert decimal point */
+        let buf = '';
+        const insertionPoint = intString.length - scale;
+        if (insertionPoint === 0) { /* Point goes right before intVal */
+            return (signum < 0 ? '-0.' : '0.') + intString;
+        } else if (insertionPoint > 0) { /* Point goes inside intVal */
+            buf = intString.slice(0, insertionPoint) + '.' + intString.slice(insertionPoint);
+            if (signum < 0) {
+                buf = '-' + buf;
+            }
+        } else { /* We must insert zeros between point and intVal */
+            buf += signum < 0 ? '-0.' : '0.';
+            for (let i = 0; i < -insertionPoint; i++) {
+                buf += '0';
+            }
+            buf += intString;
+        }
+        return buf.toString();
+    }
+
+}
+
+interface BigInterface {
+    (decimalString: string): BigDecimal;
+
+    new(decimalString: string): BigDecimal;
+}
+
+function _Big(decimalString: string): BigDecimal {
+    return BigDecimal._new(decimalString);
+}
```

Comment:
Why do we have this? Wouldn't a public constructor with `constructor(unscaledValue: BigInt, scale: number)` and a static factory method like `BigDecimal.from(decimalString: string)` would suffice?

The name `BigInterface`, or the exported `Big` does not ring a bell for me. I think they don't tell much about what they are

Reply:
This pattern is needed for exporting a constructor method that can be used with or without new keyword. `_Big` is the constructor method exported. The `BigInterface` interface is needed for type hints to be available on the exported constructor method. There may be a better method of doing this, but this worked for me.

We may rename them

Reply:
Also I can add some comments explaining this

Reply:
Why is instantiating the `BigDecimal` class without `new` keyword necessary?

Reply:
for convenience but I started think that we can remove it altogether

Reply:
Well, for me adding `new` before creating an object is not a big burden but I don't know much about conventions for libraries. Personally, my vote goes for providing construction with new only but you may try to get some opinions about it from the team